# 🚀 Project Atlas

# Portfolio Project 01

# Retail Store Sales Analytics using NumPy

**Date:** July 11, 2026

**Difficulty:** ⭐⭐⭐⭐☆

**Estimated Time:** 5–6 Hours

---

# 📌 Project Overview

A retail company has collected weekly sales data for multiple products over several weeks.

Management wants to understand product performance, weekly trends, sales consistency, and the potential impact of promotional strategies.

As a Junior Data Professional, your task is to analyze the dataset using **NumPy only** and provide meaningful business insights.

Visualization is intentionally excluded from this project to strengthen analytical thinking and array manipulation skills.

---

# 🎯 Project Objectives

By completing this project, you should be able to:

- Explore an unfamiliar dataset.
- Perform descriptive statistical analysis.
- Compare product performance.
- Analyze weekly trends.
- Filter data using Boolean masking.
- Rank products and weeks.
- Apply broadcasting for business simulations.
- Answer business questions using data.
- Document findings professionally.

---
# 📊 Dataset Description

Rows represent **Weeks**.

Columns represent **Products**.

Each value represents the weekly sales of a product.

Sales are measured in **₹ thousands**.

Example:

75

means

₹75,000 worth of sales.

---

In [93]:
import numpy as np
sales = np.array([
 [72, 81, 65, 90, 77, 84],
 [75, 79, 68, 93, 80, 86],
 [78, 85, 70, 95, 82, 88],
 [74, 83, 66, 91, 79, 85],
 [80, 88, 72, 98, 84, 90],
 [82, 91, 74,100, 86, 92],
 [79, 87, 71, 97, 83, 89],
 [81, 90, 73,101, 85, 91]
])
print("Sales : ",sales)

Sales :  [[ 72  81  65  90  77  84]
 [ 75  79  68  93  80  86]
 [ 78  85  70  95  82  88]
 [ 74  83  66  91  79  85]
 [ 80  88  72  98  84  90]
 [ 82  91  74 100  86  92]
 [ 79  87  71  97  83  89]
 [ 81  90  73 101  85  91]]


# 📖 Part 1 – Dataset Exploration

## Objective

Understand the dataset before performing any analysis.

---

## Tasks

Without using loops:

- Find the shape of the dataset.
- Find the total number of elements.
- Find the number of dimensions.
- Find the data type.
- Find the memory used by the dataset.
- Find the number of weeks.
- Find the number of products.

---

In [94]:
#Find the shape of the dataset.
print("Shape of dataset : ",np.shape(sales))
#Find the total number of elements.
print("total number of elements : ",np.size(sales))
#Find the number of dimensions.
print("no of dimensions : ",np.ndim(sales))
#Find the data type.
print("Datatype : ", sales.dtype)
#Find the memory used by the dataset.
print("memory used by the dataset : ",sales.nbytes)
#Find the number of weeks.
weeks,products = np.shape(sales)
print("Total number of weeks : ",weeks)
#Find the number of products.
print("Total number of products : ",products)

Shape of dataset :  (8, 6)
total number of elements :  48
no of dimensions :  2
Datatype :  int64
memory used by the dataset :  384
Total number of weeks :  8
Total number of products :  6


# 📖 Part 2 – Descriptive Statistics

## Objective

Generate an overall statistical summary of the dataset.

---

## Tasks

Calculate:

- Total sales
- Average sales
- Median sales
- Maximum sales
- Minimum sales
- Standard deviation
- Variance

---

## Product-wise Statistics

Calculate:

- Product-wise totals
- Product-wise averages
- Product-wise maximum
- Product-wise minimum

---

## Weekly Statistics

Calculate:

- Weekly totals
- Weekly averages
- Weekly maximum
- Weekly minimum

---

In [95]:
#Total sales
print("Total sales = ",np.sum(sales))
#Average sales
sales_average = np.mean(sales)
print("Average sales = ",sales_average)
#Median sales
print("Median sales = ",np.median(sales))
#Maximum sales
print("Max sales = ",np.max(sales))
#Minimum sales
print("Min sales = ",np.min(sales))
#Standard deviation
print("Standard Deviation = ",np.std(sales))
#Variance
print("Variance = ",np.var(sales))

Total sales =  3990
Average sales =  83.125
Median sales =  83.5
Max sales =  101
Min sales =  65
Standard Deviation =  8.764571200768087
Variance =  76.81770833333333


### Product-wise Statistics

In [96]:
#Product-wise totals
print("Product-wise totals = ",np.sum(sales,axis=0))
#Product-wise averages
product_average = np.mean(sales,axis=0)
print("Product-wise averages = ",product_average)
#Product-wise maximum
print("Product-wise maximum = ",np.max(sales,axis=0))
#Product-wise minimum
print("Product-wise minimum = ",np.min(sales,axis=0))

Product-wise totals =  [621 684 559 765 656 705]
Product-wise averages =  [77.625 85.5   69.875 95.625 82.    88.125]
Product-wise maximum =  [ 82  91  74 101  86  92]
Product-wise minimum =  [72 79 65 90 77 84]


### Weekly Statistics

In [97]:
#Weekly totals
weekly_totals = np.sum(sales,axis=1)
print("Weekly totals = ",weekly_totals)
#Weeklyaverages
weekly_average =np.mean(sales,axis=1)
print("Weekly averages = ",weekly_average)
#Weekly maximum
print("Weekly maximum = ",np.max(sales,axis=1))
#Weekly minimum
print("Weekly minimum = ",np.min(sales,axis=1))

Weekly totals =  [469 481 498 478 512 525 506 521]
Weekly averages =  [78.16666667 80.16666667 83.         79.66666667 85.33333333 87.5
 84.33333333 86.83333333]
Weekly maximum =  [ 90  93  95  91  98 100  97 101]
Weekly minimum =  [65 68 70 66 72 74 71 73]


# 📖 Part 3 – Product Performance Analysis

## Objective

Determine how each product performed.

---

## Tasks

Identify:

- Best-selling product
- Lowest-selling product
- Product ranking
- Difference between highest and lowest selling products
- Products performing above overall average
- Products performing below overall average

---

In [98]:
#Best-selling product
print("Best-selling product : ",np.argmax(np.sum(sales,axis=0)))
#Lowest-selling product
print("Lowest-selling product : ",np.argmin(np.sum(sales,axis=0)))
#Product ranking
print("Product ranking : ",np.argsort(np.sum(sales,axis=0)))
#Difference between highest and lowest selling products
sorted_products = np.sort(np.sum(sales,axis=0))
print("Difference between highest and lowest selling products : ",sorted_products[-1]-sorted_products[0])
#Products performing above overall average
print("Products performing above overall average : ",np.where(product_average>sales_average))
#Products performing below overall average
print("Products performing below overall average : ",np.where(product_average<sales_average))

Best-selling product :  3
Lowest-selling product :  2
Product ranking :  [2 0 4 1 5 3]
Difference between highest and lowest selling products :  206
Products performing above overall average :  (array([1, 3, 5]),)
Products performing below overall average :  (array([0, 2, 4]),)


# 📖 Part 4 – Weekly Performance Analysis

## Objective

Understand business performance over time.

---

## Tasks

Identify:

- Best-performing week
- Lowest-performing week
- Difference between best and worst weeks
- Weeks above average
- Weeks below average

---

In [99]:
#Best-performing week
print("Best-performing week : ",np.argmax(np.sum(sales,axis=1)))
#Lowest-performing week
print("Lowest-performing week : ",np.argmin(np.sum(sales,axis=1)))
#Difference between best and worst weeks
sorted_weeks = np.sort(np.sum(sales,axis=1))
print("Difference between best and worst weeks : ",sorted_weeks[-1]-sorted_weeks[0])
#Weeks above average
print("Weeks above average : ",np.where(weekly_average>sales_average))
#Weeks below average
print("Weeks below average : ",np.where(weekly_average<sales_average))

Best-performing week :  5
Lowest-performing week :  0
Difference between best and worst weeks :  56
Weeks above average :  (array([4, 5, 6, 7]),)
Weeks below average :  (array([0, 1, 2, 3]),)


# 📖 Part 5 – Boolean Analysis

## Objective

Filter the dataset using different business conditions.

---

## Tasks

Find:

- Sales greater than 90
- Sales below 75
- Sales between 80 and 95
- Products averaging above 82
- Weeks with total sales above the weekly average
- Values equal to exactly 100

---

In [100]:
#Sales greater than 90
print("Sales greater than 90 : ",sales[sales>90])
#Sales below 75
print("Sales below 75 : ",sales[sales<75])
#Sales between 80 and 95
print("Sales between 80 and 95 : ",sales[(sales>80)&(sales<95)])
#Products averaging above 82
print("Products averaging above 82 : ",np.where(product_average>82)[0])
#Weeks with total sales above the weekly average
print("Weeks with total sales above the weekly average : ",np.where(weekly_totals>np.mean(weekly_totals))[0])
#Values equal to exactly 100
is_100 = (sales==100)
print("Values equal to exactly 100 : ",np.sum(is_100))

Sales greater than 90 :  [ 93  95  91  98  91 100  92  97 101  91]
Sales below 75 :  [72 65 68 70 74 66 72 74 71 73]
Sales between 80 and 95 :  [81 90 84 93 86 85 82 88 83 91 85 88 84 90 82 91 86 92 87 83 89 81 90 85
 91]
Products averaging above 82 :  [1 3 5]
Weeks with total sales above the weekly average :  [4 5 6 7]
Values equal to exactly 100 :  1


# 📖 Part 6 – Sorting & Ranking

## Objective

Rank products and sales values.

---

## Tasks

Without using loops:

- Top 10 highest sales values
- Bottom 10 sales values
- Sort every product's sales
- Rank products based on total sales
- Rank weeks based on total sales

---

In [101]:
#Top 10 highest sales values
print("Top 10 highest sales values : ",np.sort(sales.flatten())[::-1][:10])
#Bottom 10 sales values
print("Bottom 10 sales values : ",np.sort(sales.flatten())[:10])
#Sort every product's sales
print("Sort every product's sales : ",np.sort(np.sum(sales,axis=0)))
#Rank products based on total sales
print("Rank products based on total sales : ",np.argsort(np.sum(sales,axis=0)))
#Rank weeks based on total sales
print("Rank weeks based on total sales : ",np.argsort(np.sum(sales,axis=1)))

Top 10 highest sales values :  [101 100  98  97  95  93  92  91  91  91]
Bottom 10 sales values :  [65 66 68 70 71 72 72 73 74 74]
Sort every product's sales :  [559 621 656 684 705 765]
Rank products based on total sales :  [2 0 4 1 5 3]
Rank weeks based on total sales :  [0 3 1 2 6 4 7 5]


# 📖 Part 7 – Business Simulations

## Objective

Simulate possible business scenarios.

---

## Scenario 1

A festival season increases all sales by **10%**.

---

## Scenario 2

Product 4 receives an additional **15% promotional boost**.

---

## Scenario 3

Returns reduce all sales by **5%**.

---

## Scenario 4

Increase Products 1 and 6 by **8%** while leaving all others unchanged.

---

In [102]:
#A festival season increases all sales by 10%
festive=sales.copy().astype(float)
festive*=1.10
print("increases all sales by 10% : ",festive)
#Product 4 receives an additional 15% promotional boost
festive[:,3]*=1.15
print("Product 4 receives an additional 15% promotional boost : ",festive)
#Returns reduce all sales by 5%
#festive*=0.95
#print("Returns reduce all sales by 5% : ",festive)
#Increase Products 1 and 6 by 8% while leaving all others unchanged.
festive[:,::5]*1.08
print("Increase Products 1 and 6 by 8% while leaving all others unchanged : ",festive)

increases all sales by 10% :  [[ 79.2  89.1  71.5  99.   84.7  92.4]
 [ 82.5  86.9  74.8 102.3  88.   94.6]
 [ 85.8  93.5  77.  104.5  90.2  96.8]
 [ 81.4  91.3  72.6 100.1  86.9  93.5]
 [ 88.   96.8  79.2 107.8  92.4  99. ]
 [ 90.2 100.1  81.4 110.   94.6 101.2]
 [ 86.9  95.7  78.1 106.7  91.3  97.9]
 [ 89.1  99.   80.3 111.1  93.5 100.1]]
Product 4 receives an additional 15% promotional boost :  [[ 79.2    89.1    71.5   113.85   84.7    92.4  ]
 [ 82.5    86.9    74.8   117.645  88.     94.6  ]
 [ 85.8    93.5    77.    120.175  90.2    96.8  ]
 [ 81.4    91.3    72.6   115.115  86.9    93.5  ]
 [ 88.     96.8    79.2   123.97   92.4    99.   ]
 [ 90.2   100.1    81.4   126.5    94.6   101.2  ]
 [ 86.9    95.7    78.1   122.705  91.3    97.9  ]
 [ 89.1    99.     80.3   127.765  93.5   100.1  ]]
Increase Products 1 and 6 by 8% while leaving all others unchanged :  [[ 79.2    89.1    71.5   113.85   84.7    92.4  ]
 [ 82.5    86.9    74.8   117.645  88.     94.6  ]
 [ 85.8    93.5   

# 📖 Part 8 – Executive Summary

Based on the analysis of the retail sales dataset, the following key business insights were identified:

- **Product 4** emerged as the best-performing product, consistently recording the highest sales across the observed weeks.
- **Product 3** showed comparatively lower sales than the other products and may require further investigation to identify opportunities for improvement.
- **Week 6** generated the highest overall revenue, indicating the strongest sales performance during the observed period.
- **Week 1** recorded the lowest overall revenue, suggesting a relatively slower start compared to subsequent weeks.

## Business Recommendations

- Maintain sufficient inventory for **Product 4** to meet customer demand.
- Analyze factors affecting **Product 3**, such as pricing, promotions, or customer preferences, to improve its performance.
- Investigate the factors that contributed to the strong performance in **Week 6** and consider replicating similar strategies in future campaigns.
- Review sales trends from **Week 1** to understand potential reasons for lower revenue and identify improvement opportunities.

# 🌟 Bonus Challenge

Without using loops:

Can you answer the following?

- Find the top three products.
- Find the top three weeks.
- Find products whose average sales exceed the median product average.
- Identify weeks where **every product** sold more than 75.
- Count how many times Product 4 achieved sales above 95.
- Calculate each product's contribution as a percentage of total sales.
- Determine which week had the smallest variation in product sales.
- Identify the most consistent product based on standard deviation.
- Simulate a 12% increase for products with average sales below the overall average.
- Create a summary table (using NumPy only) containing:
  - Product Total
  - Product Average
  - Product Maximum
  - Product Minimum
  - Product Standard Deviation

---

In [103]:
#Find the top three products.
print("top three products : ",np.argsort(np.sum(sales,axis=0))[::-1][:3])
#Find the top three weeks.
print("top three weeks",np.argsort(np.sum(sales,axis=1))[::-1][:3])
#Find products whose average sales exceed the median product average.
print("products whose average sales exceed the median product average : ",np.where(product_average>np.median(product_average))[0])
#Identify weeks where **every product** sold more than 70.
print("weeks where **every product** sold more than 70 : ",np.where(np.all((sales>70),axis=1))[0])
#Count how many times Product 4 achieved sales above 95.
prod4_95 = (sales[:,3]>95)
print("Count how many times Product 4 achieved sales above 95 : ",np.sum(prod4_95))
#Calculate each product's contribution as a percentage of total sales.
products_total = np.sum(sales,axis=0).astype(float)
total_sale = np.sum(sales)
products_contribution = products_total/total_sale
products_contribution *=100
print("each product's contribution as a percentage of total sales : ",products_contribution)
#Simulate a 12% increase for products with average sales below the overall average.
total_average = np.mean(sales)
prods_below_overall = np.where(product_average<total_average ,1.12,1.0)
new_sales = sales*prods_below_overall
print("Simulate a 12%\ increase for products with average sales below the overall average : ",new_sales)


top three products :  [3 5 1]
top three weeks [5 7 4]
products whose average sales exceed the median product average :  [1 3 5]
weeks where **every product** sold more than 70 :  [4 5 6 7]
Count how many times Product 4 achieved sales above 95 :  4
each product's contribution as a percentage of total sales :  [15.56390977 17.14285714 14.01002506 19.17293233 16.44110276 17.66917293]
Simulate a 12%\ increase for products with average sales below the overall average :  [[ 80.64  81.    72.8   90.    86.24  84.  ]
 [ 84.    79.    76.16  93.    89.6   86.  ]
 [ 87.36  85.    78.4   95.    91.84  88.  ]
 [ 82.88  83.    73.92  91.    88.48  85.  ]
 [ 89.6   88.    80.64  98.    94.08  90.  ]
 [ 91.84  91.    82.88 100.    96.32  92.  ]
 [ 88.48  87.    79.52  97.    92.96  89.  ]
 [ 90.72  90.    81.76 101.    95.2   91.  ]]


In [104]:
#Create a summary table (using NumPy only) containing:
#  - Product Total
#  - Product Average
#  - Product Maximum
#  - Product Minimum
#  - Product Standard Deviation

products_max = np.max(sales,axis=0)
products_min = np.min(sales,axis=0)
products_std = np.std(sales,axis=0)

s_total = np.transpose(products_total)
s_avg = np.transpose(product_average)
s_max = np.transpose(products_max)
s_min = np.transpose(products_min)
s_std = np.transpose(products_std)

table = np.column_stack((s_total,s_avg,s_max,s_min,s_std))
print(table)


[[621.          77.625       82.          72.           3.35177192]
 [684.          85.5         91.          79.           4.        ]
 [559.          69.875       74.          65.           3.05930956]
 [765.          95.625      101.          90.           3.80583434]
 [656.          82.          86.          77.           2.91547595]
 [705.          88.125       92.          84.           2.7128168 ]]


# 🏁 Project Conclusion

## 📚 What I Learned

This project helped me transition from learning individual NumPy functions to solving a complete business problem using NumPy. Instead of focusing on isolated operations, I combined multiple concepts to analyze a realistic retail sales dataset and derive meaningful business insights.

Throughout the project, I strengthened my understanding of:

* Data exploration using NumPy arrays
* Descriptive statistical analysis
* Product-wise and week-wise performance analysis
* Boolean masking for business conditions
* Broadcasting for business simulations
* Sorting and ranking data
* Aggregation functions with `axis=0` and `axis=1`
* Building summary tables using only NumPy

---

## 💡 Biggest Challenges

### 1. Simulating a 12% Increase

One of the more challenging parts was applying a **12% sales increase only to products whose average sales were below the overall average**.

```python
total_average = np.mean(sales)

prods_below_overall = np.where(product_average < total_average, 1.12, 1.0)

new_sales = sales * prods_below_overall
```

### How it works

* `np.mean(sales)` calculates the overall average of the dataset.
* `product_average` stores the average sales for each product.
* `np.where()` creates a multiplier array:

  * `1.12` for products below the overall average.
  * `1.00` for all other products.
* NumPy broadcasting automatically applies these multipliers to each product column without using loops.

**Concepts Used**

* Aggregation
* Boolean masking
* Broadcasting
* Vectorized operations

---

### 2. Counting Product 4 Sales Above 95

```python
prod4_95 = (sales[:,3] > 95)

np.sum(prod4_95)
```

### How it works

* `sales[:,3]` selects all rows from Product 4.
* The comparison `> 95` creates a Boolean array.
* `True` values represent weeks where Product 4 sold more than 95.
* `np.sum()` counts the number of `True` values because NumPy treats `True` as `1`.

**Concepts Used**

* Column selection
* Boolean masking
* Aggregation

---

### 3. Identifying Weeks Where Every Product Sold Above 70

```python
np.where(np.all(sales > 70, axis=1))[0]
```

### How it works

* `sales > 70` creates a Boolean matrix.
* `np.all(..., axis=1)` checks every row.
* A row becomes `True` only if **every product** satisfies the condition.
* `np.where()` returns the indices of those weeks.

**Concepts Used**

* Boolean masking
* Logical aggregation
* Row-wise analysis

---

### 4. Creating a Product Summary Table

This was another concept that initially required some thought.

The goal was to create a table containing:

* Product Total
* Product Average
* Product Maximum
* Product Minimum
* Product Standard Deviation

Instead of using Python loops, multiple statistical arrays were calculated and combined into a single summary table using NumPy's column-wise stacking functionality.

Example output:

| Product | Total | Average | Maximum | Minimum | Standard Deviation |
| ------- | ----: | ------: | ------: | ------: | -----------------: |
| P1      |   621 |   77.63 |      82 |      72 |               3.35 |
| P2      |   684 |   85.50 |      91 |      79 |               4.00 |
| P3      |   559 |   69.88 |      74 |      65 |               3.06 |
| P4      |   765 |   95.63 |     101 |      90 |               3.81 |
| P5      |   656 |   82.00 |      86 |      77 |               2.92 |
| P6      |   705 |   88.13 |      92 |      84 |               2.71 |

This exercise reinforced how NumPy can combine multiple statistical calculations into a structured analytical summary.

---

## 🚀 NumPy Concepts Applied

* Array Creation
* Array Properties
* Indexing & Slicing
* Fancy Indexing
* Reshaping
* Aggregation Functions
* Axis Operations
* Broadcasting
* Vectorized Operations
* Boolean Masking
* Searching
* Sorting
* Statistical Analysis
* Business Simulations
* Summary Table Construction

---

## 📈 Business Insights

* Product 4 consistently recorded the highest sales and was the best-performing product.
* Product 3 showed comparatively lower sales and could benefit from further analysis or promotional strategies.
* Week 6 generated the highest overall revenue.
* Week 1 recorded the lowest overall sales.
* Promotional simulations demonstrated how broadcasting can be used to model business scenarios efficiently.
* Statistical summaries helped identify product consistency and performance variation.

---

## 🔮 Future Improvements

As I continue progressing through Project Atlas, I plan to enhance this project by:

* Reading data from CSV files using Pandas.
* Cleaning and preprocessing real-world datasets.
* Creating visualizations using Matplotlib and Seaborn.
* Building interactive dashboards.
* Applying Machine Learning models for sales forecasting.
* Comparing multiple datasets over longer time periods.

---

# ✅ Project Status

**Portfolio Project 01 Successfully Completed**

Part of **Project Atlas** — a structured learning journey covering Python, Data Analytics, Machine Learning, and Artificial Intelligence.
